src/nnvm/low_precision_pass.cc

```diff
@@ -0,0 +1,265 @@
+    const auto &e = inputs[i];
+    NodePtr mirror_node = mirror_map.at(e.node.get());
+    NodeEntry mirror_entry = NodeEntry{mirror_node, e.index, e.version};
+    node->inputs.emplace_back(mirror_entry);
+    }
+    node->attrs.dict["num_outputs"] = std::to_string(inputs.size());
+    node->op()->attr_parser(&(node->attrs));
+    for (uint32_t i = 0; i < inputs.size(); ++i) {
+    const auto &e = inputs[i];
+    curr_node->inputs.emplace_back(
+        NodeEntry{node, static_cast<uint32_t>(i), e.version});
+    }
+    return;
+}
+
+bool CheckConditionalFP32(
+    const std::unordered_map<
+        std::string, std::unordered_map<std::string, std::vector<std::string>>>
+        &conditional_fp32_ops,
+    const std::unordered_set<std::string> &excluded_syms, NodePtr node) {
+  if (node->is_variable() || (excluded_syms.count(node->attrs.name) > 0) ||
+      conditional_fp32_ops.count(node->op()->name) == 0) {
+    return false;
+  } else {
+    // Iterate through all conditional ops
+    auto it = conditional_fp32_ops.find(node->op()->name);
+    if (it != conditional_fp32_ops.end()) {
+      auto it_params = it->second;
+      // For each param name, iterate through param values to check
+      // if the provided param name is equal to any of the values
+      for (auto it_param = it_params.begin(); it_param != it_params.end();
+           it_param++) {
+        auto param_key = node->attrs.dict.find(it_param->first);
+        if (param_key != node->attrs.dict.end()) {
+          auto it_param_vals = it_param->second;
+          if (std::find(it_param_vals.begin(), it_param_vals.end(),
+                        param_key->second) != it_param_vals.end()) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+}
+
+Graph ReducePrecision(Graph &&src) {
+  const auto target_dtype_ops =
+      src.GetAttr<std::unordered_set<std::string>>("target_dtype_ops");
+  const auto fp32_ops =
+      src.GetAttr<std::unordered_set<std::string>>("fp32_ops");
+  const auto widest_dtype_ops =
+      src.GetAttr<std::unordered_set<std::string>>("widest_dtype_ops");
+  const auto target_dtype = src.GetAttr<int>("target_dtype");
+  const auto excluded_syms = src.GetAttr<std::unordered_set<std::string>>("excluded_syms");
+  const auto conditional_fp32_ops = src.GetAttr<std::unordered_map<
+      std::string, std::unordered_map<std::string, std::vector<std::string>>>>(
+      "conditional_fp32_ops");
+
+  CHECK(target_dtype == mshadow::kFloat16)
+      << "Only float16 target_dtype is supported yet";
+
+  // Additional data structures to share common cast node inputs among different nodes
+  std::unordered_map<Node *, NodePtr> mirror_map;
+  nnvm::NodeEntryMap<NodeEntry> mirror_fp32_map;
+  nnvm::NodeEntryMap<NodeEntry> mirror_target_dtype_map;
+
+  // Visit nodes in a topologically sorted order
+  DFSVisit(src.outputs, [&](const NodePtr &node) {
+    NodePtr new_node = Node::Create();
```

Comment:
can't you pass node directly to Create()?  it's forwarding args so it should call the copy ctor.

Reply:
not sure about this. can you elaborate with code example.

Reply:
Why not? 

```
NodePtr new_node = Node::Create(*node);  

```

Reply:
yes you are right. my editor was getting confused by num of arguments probably because of std::forward. i have fixed this now.